larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerESTAREnergy_tool.cc

```diff
@@ -0,0 +1,293 @@
+
+      unsigned int planeNumHits = hits.size();
+        
+      //Calculate the Energy
+      double Energy = CalculateEnergy(clockData, detProp, hits, plane);
+
+      if(Energy > 0){    
+        ShowerESTAREnergy.at(plane) = Energy;
+      }
+
+      if(planeNumHits > bestPlaneNumHits){
+        bestPlane        = plane;
+        bestPlaneNumHits = planeNumHits;
+      }
+      if(fVerbose){
+        std::cout << "plane: " << plane
+                  << "  hits: " << hits.size()
+                  << "  energy: " << Energy << std::endl;  
+      }
+
+    }
+
+    ShowerEleHolder.SetElement(ShowerESTAREnergy,ShowerESTAREnergyError,fShowerEnergyOutputLabel);
+
+    // Only set the best plane if it has some hits in it
+    if (bestPlane < fGeom->Nplanes()){
+      // Need to cast as an int for legacy default of -999
+      // have to define a new variable as we pass-by-reference when filling
+      int bestPlaneVal(bestPlane);
+      ShowerEleHolder.SetElement(bestPlaneVal, fShowerBestPlaneOutputLabel);
+      if(fVerbose){
+        std::cout << "best plane: " << bestPlane << ",  best plane number of hit: " << bestPlaneNumHits << std::endl;
+      }
+    }
+     
+    return 0;
+  }
+
+
+  // function to calculate the reco energy
+  double ShowerESTAREnergy::CalculateEnergy(const detinfo::DetectorClocksData& clockData,
+         const detinfo::DetectorPropertiesData& detProp,
+         const std::vector<art::Ptr<recob::Hit> >& hits, const geo::PlaneID::PlaneID_t plane){
+   
+    double totalEnergy = 0;
+    double efield = 0; 
+     
+    // Loop over the hits
+    for(auto const& h : hits){
+      // If we're correcting for SCE, check if the hit is one that was we found to have an associated SP
+      if(fSCECorrectEField){
+        auto it = std::find_if(sp_hits_tuple.begin(), sp_hits_tuple.end(), [&](const std::tuple<unsigned int, double>& e){
+        return std::get<0>(e) == h.key();
+      });
+      
+      // do a per hit SCE correction since we have associated space points
+      if(it != sp_hits_tuple.end()){
+        efield = std::get<1>(*it);
+      }
+
+      // no associated space points :( so correct these hits using the shower centre position 
+      else{
+        efield = localEfield_cweighted;
+      }
+    }
+
+    else{
+      efield = nominal_Efield; 
+    }
+
+    double totalCharge = (h)->Integral() * fCalorimetryAlg.LifetimeCorrection(clockData, detProp, (h)->PeakTime()); // obtain charge and correct for lifetime
+    double nElectrons = fCalorimetryAlg.ElectronsFromADCArea(totalCharge, plane);
+    if(nElectrons !=0 && t_estar->Interpolate(nElectrons, efield) <= 0){
+      mf::LogWarning("ShowerESTAREnergy") << "Interpolation is returning an energy of 0 even though we do have some collected charge. " 
+                                          << "Probably due to an out of bounds entry for the charge or the efield." << std::endl; 
+    }
+    totalEnergy += t_estar->Interpolate(nElectrons, efield);
```

Comment:
It looks like the negative energy is warned about but still added on.  Maybe it should be set to 0 in this case?